### PR TITLE
trie example: one for usage and one for memory consumption test.

### PR DESCRIPTION
**Pull request by drmingdrmer**
# Description

Two trie examples I promised yesterday.

## Type of change

Please delete options that are not relevant.

- [x] Refactoring
- [x] Document changes
- [x] Test changes

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: trie/example_mem_usage_test.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+func makecopies(ncopy int, keys []string, offsets []uint16) []*trie.SlimTrie {
+	copys := make([]*trie.SlimTrie, ncopy)
+
+	for i := 0; i < len(copys); i++ {
+		copys[i], _ = trie.NewSlimTrie(marshal.U16{}, keys, offsets)
+	}
+	return copys
+}
+
+func Example_memoryUsage() {
```
Comment (lishulong) [inline trie/example_mem_usage_test.go]: 有下划线是生成example文档的工具的要求吗？

Comment (drmingdrmer) [reply trie/example_mem_usage_test.go]: 对. `Example<type or func name>_<name of example>`
ExampleSlimTrie, ExampleSlimTrie_get 酱

File: trie/example_mem_usage_test.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/openacid/slim/marshal"
+	"github.com/openacid/slim/trie"
+)
+
+var (
+	m = marshal.String16{}
+)
+
+func makecopies(ncopy int, keys []string, offsets []uint16) []*trie.SlimTrie {
+	copys := make([]*trie.SlimTrie, ncopy)
```
Comment (lishulong) [inline trie/example_mem_usage_test.go]: copies？

File: trie/example_mem_usage_test.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+	"github.com/openacid/slim/marshal"
+	"github.com/openacid/slim/trie"
+)
+
+var (
+	m = marshal.String16{}
+)
+
+func makecopies(ncopy int, keys []string, offsets []uint16) []*trie.SlimTrie {
```
Comment (lishulong) [inline trie/example_mem_usage_test.go]: makeCopies，怎么样？

File: trie/example_test.go [hunk 4]
```diff
@@ -1,5 +1,66 @@
...
+	kv := strings.Split(string(d)[offset:], ",")[0:2]
+	if kv[0] == key {
+		return kv[1], true
+	}
+	return "", false
+}
+
 func Example() {
 
+	// SlimTrie is born a memory efficient data type as an index.
```
Comment (lishulong) [inline trie/example_test.go]: Usually used as an index, SlimTrie is designed to be a memory efficient data type.

怎么样？

File: trie/example_test.go [hunk 5]
```diff
@@ -1,5 +1,66 @@
...
+	//
+	// In this example we shows that how to index external data(in memory or on
```
Comment (lishulong) [inline trie/example_test.go]: In this example, we show the way to accelerate accessing to external data (in memory or on disk) by indexing them in a SlimTrie: 

怎么样？

File: trie/example_test.go [hunk 6]
```diff
@@ -1,5 +1,66 @@
...
+	// disk), to accelerate accessing:
+	//
+	// `data` is a sample of some unindexed data. In our example it is a comma
+	// seperated key value series.
+	//
+	// In order to let SlimTrie to read data, `data` should implement
```
Comment (lishulong) [inline trie/example_test.go]: In order to make SlimTrie read data 
?

Comment (lishulong) [reply trie/example_test.go]: `data` should have a `Read` method.
?

File: trie/example_test.go [hunk 7]
```diff
@@ -1,5 +1,66 @@
...
+	// a `Read` method:
+	//     Read(offset int64, key string) (string, bool)
+	data := Data("Aaron,1,Agatha,1,Al,2,Albert,3,Alexander,5,Alison,8")
+
+	// `index` is a prebuilt index of a key and its offset in `data`.
```
Comment (lishulong) [inline trie/example_test.go]: `index` is a prebuilt index that stores each key with its offset in `data` accordingly.

这样如何？

File: trie/example_test.go (hunk 6 above)
Comment (drmingdrmer) [reply trie/example_test.go]: > In order to make SlimTrie read data
> ?

感觉这里用let好点...make感觉有情绪...

Comment (drmingdrmer) [reply trie/example_test.go]: In order to let SlimTrie be able to read data

酱?

File: trie/example_mem_usage_test.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		avgIdxLen,
+		100-ratio,
+	)
+
+	for _, cc := range copys {
+		_ = cc.Children
+	}
+}
+
+func readRss() int64 {
```
Comment (pengsven) [inline trie/example_mem_usage_test.go]: array32也用到了，之后考虑放在一个共用的地方?

Comment (drmingdrmer) [reply trie/example_mem_usage_test.go]: 只有测试用到...而且...这个函数好像有点问题...有时会拿到负的变化, stat里还有几个alloc相关的变量还没仔细看. 等整明白了在看是否独立出来

**Pull request merged**